design/one-pager-terrajet-metadata-extraction.md

```diff
@@ -0,0 +1,272 @@
+                update: '- (Defaults to 30 minutes) Used when updating the Analysis Services Server.'
+        # import statement scraped from the Terraform registry, if available
+        # Can be used for advanced purposes, such as constructing resource config "ExternalName.GetIDFn" functions, etc.
+        importStatements:
+            - terraform import azurerm_analysis_services_server.server /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup1/providers/Microsoft.AnalysisServices/servers/server1
+```
+
+Another alternative could be to have qualified names under the `argumentDocs`
+with a flat hierarchy, e.g., instead of a nested `ipv4_firewall_rule` block
+represented as a map, we could have its block parameters qualified with the
+configuration block name (`ipv4_firewall_rule.range_start`,
+`ipv4_firewall_rule.range_end`, etc.) Then `argumentDocs` would become a simple
+`map[string]string`.
+
+Another alternative could be to have per-resource YAML metadata files, i.e.,
+instead of the `resources` map in a single file, we could have each of its keys
+(and associated metadata) stored in a resource specific YAML-formatted file.
+These resource specific files could each be named as `<Terraform resource
+type>.yaml`, e.g., `azurerm_analysis_services_server.yaml`. 
+
+### Metadata scrapers
+Although not validated on all of available Terraform providers, at least, the
+big three Terraform providers ([terraform-provider-aws],
+[terraform-provider-azurerm] and [terraform-provider-google]) all have Terraform
+registry content in their respective repositories and use markdown documents
+with a common structure. Our assumption is that Terraform registry website is
+also generated using these markdown files:
+- For `terraform-provider-aws`:
+  https://github.com/hashicorp/terraform-provider-aws/tree/main/website/docs/r
+- For `terraform-provider-azurerm`:
+  https://github.com/hashicorp/terraform-provider-azurerm/tree/main/website/docs/r
+- For `terraform-provider-google`:
+  https://github.com/hashicorp/terraform-provider-google/tree/main/website/docs/r
+
+Thus a common metadata scraper implementation can extract metadata from these
+well-formatted per-resource markdown documents. Any spotted errors can then
+potentially be corrected manually in the scraped YAML metadata document.
+Scrapers can optionally be chained: If desired, another scraper can append
+example HCL configurations read from a different source (such as the `examples`
+folder found in some of the Terraform native provider repositories as discussed
+above).
+
+For most Terraform native providers, we anticipate that Terraform registry
+scrapers will **not** run on HTTP, as the resource markdown files are part of
+their corresponding provider repositories. They can just read those markdowns
+from a pointed directory in the local filesystem, which is specified as a
```

Comment:
Do you mean it will fetch the `.md` files from the Go module directory or the repositories will be cloned manually to a folder in the filesystem?

Reply:
If it's the latter, I wonder if we can go with HTTP just to have it fully automated.

Reply:
I think we can selectively git-checkout the repo path that contains the markdown files, as part of a `Makefile` target. I have also extended the discussion on alternative transports for the scrapers.